### PR TITLE
[PAN-2090] ESLint — Add new rules to enforce our existing coding style

### PR DESCRIPTION
**Pull request by german-panoply**
Jira Task: https://panoply.atlassian.net/browse/PAN-2090

## Description
- Added new enforce our existing coding style
- Preserved all the exceptions we already had
- Refactored the config for better readability
- Enabled the linter to work on this repo too

> **Note:** I'll double check everything a few times before merging anything.

> **Note:** In the future, we might want to get rid of the exceptions as well.

## Related PRs
- https://github.com/panoplyio/ui-components/pull/292

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -8,127 +8,192 @@ module.exports = {
...
 
         // allow usage of `arguments` instead of only rest params (...args)
-        'prefer-rest-params': [ 0 ],
+        'prefer-rest-params': 'off',
 
         // allow the usage of `apply()` instead of only spread operator
-        'prefer-spread': [ 0 ],
+        'prefer-spread': 'off',
 
         // allow usage of `this` keyword outside of classes/class-like objects.
```
Comment (OmriGLevy) [inline index.js]: Why are we allowing it  ?

File: package.json [hunk 2]
```diff
@@ -23,7 +24,7 @@
   },
   "homepage": "https://github.com/panoplyio/eslint-config-panoplioy#readme",
   "dependencies": {
-    "eslint-config-google": "^0.14.0",
```
Comment (OmriGLevy) [inline package.json:26]: We should remove the style guide from our Confluence as well, I remember we refer to it.

File: index.js (hunk 1 above)
Comment (german-panoply) [reply index.js]: I'm keeping all these exceptions as for now because it would be overwhelming to revisit them all at once.

**Review by OmriGLevy — APPROVED** [review]
Looks good to me, just make sure you delete the confluence file that refers to google's style sheet.

File: package.json (hunk 2 above)
Comment (german-panoply) [reply package.json:26]: @OmriGLevy can't find any mention of this, you mean this doc? https://panoply.atlassian.net/wiki/spaces/RD/pages/1014858095/Coding+Style+Conventions

**Comment by german-panoply** [comment]
> Looks good to me, just make sure you delete the confluence file that refers to google's style sheet.

Thank you!

This repo is configured to be approved by the top hierarchy since it might affect others (CC: @yonivy)

File: index.js [hunk 3]
```diff
@@ -8,127 +8,192 @@ module.exports = {
...
-        sourceType: 'module'
+        sourceType: 'module',
     },
     rules: {
-        'new-cap': [ 2, {
-            capIsNewExceptions: [ 'Polymer', 'Router' ]
+        // Possible Errors
+        'no-console': 'error',
+        'no-import-assign': 'error',
+        'no-inner-declarations': 'off',
```
Comment (yonivy) [inline index.js]: Where in the code does this break?

File: index.js [hunk 4]
```diff
@@ -8,127 +8,192 @@ module.exports = {
...
-        'node': true,
-        'browser': true,
-        'es6': true
+        node: true,
+        browser: true,
+        es6: true,
     },
-    extends: 'google'
+    extends: [
+        'eslint:recommended',
```
Comment (yonivy) [inline index.js]: 👏

File: index.js [hunk 5]
```diff
@@ -8,127 +8,192 @@ module.exports = {
...
+        'no-use-before-define': 'off',
         // prevent debugger statements in the code
-        'no-debugger': [ 2 ],
+        'no-debugger': 'error',
 
-        'prefer-const': [0],
+        'prefer-const': 'off',
 
-        'prefer-promise-reject-errors': [0],
+        'prefer-promise-reject-errors': 'off',
```
Comment (yonivy) [inline index.js]: This is better on.

File: index.js [hunk 6]
```diff
@@ -8,127 +8,192 @@ module.exports = {
...
+        'keyword-spacing': ['error', { before: true }],
+        'no-multi-assign': 'error',
+        'no-trailing-spaces': 'error',
+        'no-unneeded-ternary': 'error',
+        'jsx-quotes': ['error', 'prefer-double'],
+        'quote-props': ['error', 'consistent-as-needed'],
+        'spaced-comment': ['error', 'always', { block: { balanced: true } }],
+        'multiline-comment-style': ['error', 'separate-lines'],
+        'prefer-template': 'off',
+        'class-methods-use-this': 'off',
```
Comment (yonivy) [inline index.js]: This is better on.

File: index.js [hunk 7]
```diff
@@ -8,127 +8,192 @@ module.exports = {
...
-        'eqeqeq': ["error", "always", { "null": "ignore" }],
+        'eqeqeq': ['error', 'always', { null: 'ignore' }],
 
         // jsdoc is required but doesn't have to be valid
         // this is in order to enforce some comments on classes and method
         // without requiring a lot of redundant boilerplate
-        'require-jsdoc': [ 2 ],
-        'valid-jsdoc': [ 0 ],
+        'require-jsdoc': 'error',
+        'valid-jsdoc': 'off',
```
Comment (yonivy) [inline index.js]: This was deprecated in v5.10.0 so why is this here?

See https://eslint.org/docs/latest/rules/valid-jsdoc

File: index.js [hunk 8]
```diff
@@ -8,127 +8,192 @@ module.exports = {
...
-        'padded-blocks': [ 0 ],
+        'padded-blocks': 'off',
 
         // allow TODO and FIXME comments
-        'no-warning-comments' : [ 0 ],
+        'no-warning-comments': 'off',
 
         // allow usage of the `var` keyword
-        'no-var': [ 0 ],
+        'no-var': 'off',
```
Comment (yonivy) [inline index.js]: This is better on.

File: index.js [hunk 9]
```diff
@@ -8,127 +8,192 @@ module.exports = {
...
-        'prefer-rest-params': [ 0 ],
+        'prefer-rest-params': 'off',
 
         // allow the usage of `apply()` instead of only spread operator
-        'prefer-spread': [ 0 ],
+        'prefer-spread': 'off',
 
         // allow usage of `this` keyword outside of classes/class-like objects.
-        'no-invalid-this': [ 0 ],
+        'no-invalid-this': 'off',
```
Comment (yonivy) [inline index.js]: This is better on.

File: index.js [hunk 10]
```diff
@@ -8,127 +8,192 @@ module.exports = {
...
         // to generate cleaner code where the core functionality is at the top
         // and helpers are at the bottom. Instead of vice-versa.
-        'no-use-before-define': [ 0 ],
+        'no-use-before-define': 'off',
         // prevent debugger statements in the code
-        'no-debugger': [ 2 ],
+        'no-debugger': 'error',
 
-        'prefer-const': [0],
+        'prefer-const': 'off',
```
Comment (yonivy) [inline index.js]: This is better on.

**Review by yonivy — COMMENTED** [review]
Great job on keeping our linter updated. Only comments and a recommendation from my end.

- Checkout all my `This is better on.` comments, I believe those should be 'on' and we should override the rules only in the `panoply` repo which is the cause for the 'off' setting.
- It'd be even better to exclude these rules only in the specific files that violate them. Put the exclusions at the top of the file, that will prevent these exclusions from allowing bad practices and different styles from spreading in the panoply codebase.

File: index.js (hunk 3 above)
Comment (german-panoply) [reply index.js]: This can be also "on", it doesn't break anything, as I can see.

**Comment by german-panoply** [comment]
@yonivy Enabled some rules based on your suggested. 

Plus:
- Added sub-config for react apps, will be used as `panoplyio/react`.
- Because of this, split common rules into a separate `common.js` file.
- Moved stuff that relevant only for the panoply repo out of this repo.

**Review by yonivy — APPROVED** [review]
